src/main/java/uk/gov/ons/ctp/common/cloud/TestCloudDataStore.java

```diff
@@ -0,0 +1,104 @@
+import com.google.cloud.firestore.QuerySnapshot;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import javax.annotation.PostConstruct;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import uk.gov.ons.ctp.common.error.CTPException;
+import uk.gov.ons.ctp.common.firestore.FirestoreWait;
+
+/** Access the Cloud data store for testing purposes. */
+@Component
+public class TestCloudDataStore implements CloudDataStore {
+  private static final int DELETION_BATCH_SIZE = 100;
+
+  @Autowired private FirestoreDataStore dataStore;
+
+  private Firestore firestore;
+
+  @PostConstruct
+  public void create() {
+    firestore = FirestoreOptions.getDefaultInstance().getService();
+  }
+
+  public long deleteCollection(String schema) {
+    CollectionReference collection = firestore.collection(schema);
+
+    long totalDeleted = 0;
+    int numDeletedInBatch = 0;
+    do {
+      numDeletedInBatch = deleteBatch(collection);
+      totalDeleted += numDeletedInBatch;
+    } while (numDeletedInBatch >= DELETION_BATCH_SIZE);
+
+    return totalDeleted;
+  }
+
+  @Override
+  public void deleteObject(String schema, String key) throws CTPException {
+    dataStore.deleteObject(schema, key);
+  }
+
+  @Override
+  public Set<String> getCollectionNames() {
+    return dataStore.getCollectionNames();
+  }
+
+  @Override
+  public <T> Optional<T> retrieveObject(Class<T> target, String schema, String key)
+      throws CTPException {
+    return dataStore.retrieveObject(target, schema, key);
+  }
+
+  @Override
+  public <T> List<T> search(Class<T> target, String schema, String[] fieldPath, String searchValue)
+      throws CTPException {
+    return dataStore.search(target, schema, fieldPath, searchValue);
+  }
+
+  @Override
+  public void storeObject(String schema, String key, Object value) throws CTPException {
+    try {
+      dataStore.storeObject(schema, key, value);
+    } catch (DataStoreContentionException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public boolean waitForObject(String collection, String key, long timeoutMillis)
+      throws CTPException {
+
+    if (collection == null || key == null) {
+      throw new IllegalArgumentException("collection and key must be provided");
+    }
+
+    FirestoreWait firestore =
+        FirestoreWait.builder().collection(collection).key(key).timeout(timeoutMillis).build();
+
+    return firestore.waitForObject() != null;
+  }
+
+  // there is no firestore method to delete a collection, so we delete in batches.
+  private int deleteBatch(CollectionReference collection) {
+    int deleted = 0;
+    try {
+      ApiFuture<QuerySnapshot> future = collection.limit(DELETION_BATCH_SIZE).get();
+      List<QueryDocumentSnapshot> documents = future.get().getDocuments();
+      for (QueryDocumentSnapshot document : documents) {
+        document.getReference().delete();
```

Comment:
The delete call is actually an asynchronous operation. You should really wait until the object has been deleted.

In the unlikely event that a delete failed the test wouldn't currently know about it. Also, the collection is almost certainly not empty when the method returns, which opens up the possibility of race conditions.

Suggest either:

- delegate the delete to dataStore.deleteObject()
- Issue all the deletes and then wait for them all to complete (probably faster but a bit more complex)

Reply:
Option 2 has been chosen.